File: curriculum/algos/BubbleSort.java
final class BubbleSort {
static void bubbleSort(int[] values) {
for (int boundary = values.length; boundary > 1; --boundary) {
boolean changed = false;
for (int index = 1; index < boundary; ++index) {
if (values[index] < values[index - 1]) {
int temporary = values[index - 1];
values[index - 1] = values[index];
values[index] = temporary;
changed = true;
}
}
if (!changed) break;
}
}
}

File: curriculum/algos/bubble_sort.cpp
// Generic bubble sort using adjacent iterator operations.
#include <functional>
#include <iterator>
#include <utility>
template<typename Iterator, typename Compare = std::less<>>
void bubble_sort(Iterator first, Iterator last, Compare compare = {}) {
if (first == last) return;
for (Iterator boundary = last; boundary != first; --boundary) {
bool changed = false;
Iterator left = first;
Iterator right = std::next(left);
while (right != boundary) {
if (compare(*right, *left)) {
std::iter_swap(left, right);
changed = true;
}
++left;
++right;
}
if (!changed) break;
}
}
